test/client/records-queue.spec.js

```diff
@@ -0,0 +1,526 @@
+                    } );
+                } )
+                .then( done, done );
+        } );
+
+        it( 'updates an autosave draft record with files', done => {
+            recordA.files = files.slice();
+
+            records.updateAutoSavedRecord( recordA )
+                .then( () => {
+                    return records.getAutoSavedRecord();
+                } )
+                .then( record => {
+                    expect( record.draft ).to.equal( true );
+                    expect( record.xml ).to.equal( record.xml );
+                    expect( record.files.length ).to.equal( files.length );
+
+                    for ( const [ index, file ] of files.entries() ) {
+                        const updated = record.files[index];
+
+                        expect( updated.name ).to.equal( file.name );
+                        expect( updated.item ).to.to.be.an.instanceof( Blob );
+                    }
+                } )
+                .then( done, done );
+        } );
+
+        it( 'creates a record with the current autosaved record\'s files', done => {
+            const autoSavedUpdate = Object.assign( {}, recordA, {
+                files: files.slice(),
+            } );
+
+            records.updateAutoSavedRecord( autoSavedUpdate )
+                .then( () => {
+                    return records.save( 'set', recordA );
+                } )
+                .then( () => {
+                    return store.record.get( instanceIdA );
+                } )
+                .then( record => {
+                    expect( record.files.length ).to.equal( files.length );
+
+                    for ( const [ index, file ] of files.entries() ) {
+                        const updated = record.files[index];
+
+                        expect( updated.name ).to.equal( file.name );
+                        expect( updated.item ).to.to.be.an.instanceof( Blob );
+                    }
+                } )
+                .then( done, done );
+        } );
+
+        it( 'updates a record', done => {
+            const update = {
+                draft: false,
+                enketoId,
+                instanceId: instanceIdA,
+                name: 'name A updated',
+                xml: '<model><updated/></model>'
+            };
+            const payload = Object.assign( {}, update );
+
+            records.save( 'set', recordA )
+                .then( () => {
+                    return records.save( 'update', payload );
+                } )
+                .then( () => {
+                    return store.record.get( instanceIdA );
+                } )
+                .then( record => {
+                    Object.entries( update ).forEach( ( [ key, value ] ) => {
+                        expect( record[key] ).to.deep.equal( value );
+                    } );
+                } )
+                .then( done, done );
+        } );
+
+        it( 'creates a last-saved record when creating a record', done => {
```

Comment:
I like this suggestion (and similar ones) a lot! And it's very much in the BDD spirit of the test framework APIs. It's generally more common in JS/TS to pair test modules to a module of the same name, which can make discovery a little easier, but I really do prefer the feature-centered approach. I'll look at that in the morning.

Reply:
I'm glad to see the concept resonates. Of course, the change doesn't necessarily have to happen in this PR unless it's straightforward and does end up adding clarity. I mostly want it out there as something to consider as more tests get added.

> which can make discovery a little easier

Depending on which "direction" you're coming from! Like if you see some last-saved behavior in the form cache or just know about the feature and want to see corresponding tests, these records-queue tests might be hard to find. The biggest challenge I have with a purely module-based structure is that it's hard to reason about conceptual coverage (not line coverage) because an edge case could be tested in a number of places.